src/qiboml/operations/expectation.py

```diff
@@ -0,0 +1,111 @@
+from qibo.config import raise_error
+
+from qiboml.operations.differentiation import parameter_shift
+
+
+def expectation(
+    observable: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    initial_state: Optional[Union[List, qibo.Circuit]] = None,
+    nshots: int = 1000,
+    backend: str = "qibojit",
+    differentiation_rule: Optional[callable] = parameter_shift,
+):
+    """
+    Compute the expectation value of ``observable`` over the state obtained by
+    executing ``circuit`` starting from ``initial_state``. The final state is
+    reconstructed from ``nshots`` execution of ``circuit`` on the selected ``backend``.
+    In addition, a differentiation rule can be set, which is going to be integrated
+    within the used high-level framework. For example, if TensorFlow is used
+    in the user code and one parameter shift rule is selected as differentiation
+    rule, the expectation value is computed informing the TensorFlow graph to
+    use as gradient the output of the parameter shift rule executed on the selected
+    backend.
+
+    Args:
+        observable (qibo.Hamiltonian): the observable whose expectation value has
+            to be computed.
+        circuit (qibo.Circuit): quantum circuit returning the final state over which
+            the expectation value of ``observable`` is computed.
+        initial_state (Optional[Union[List, qibo.Circuit]]): initial state on which
+            the quantum circuit is applied.
+        nshots (int): number of times the quantum circuit is executed. Increasing
+            the number of shots will reduce the variance of the estimated expectation
+            value while increasing the computational cost of the operation.
+        backend (str): backend on which the circuit is executed. This same backend
+            is used if the chosen differentiation rule makes use of expectation
+            values.
+        differentiation_rule (Optional[callable]): the chosen differentiation
+            rule. It can be selected among the methods implemented in
+            ``qiboml.differentiation``.
+    """
+
+    # read the frontend user choice
+    frontend = observable.backend
+
+    if isinstance(frontend, TensorflowBackend):
+        return expectation_with_tf(
+            observable=observable,
+            circuit=circuit,
+            initial_state=initial_state,
+            nshots=nshots,
+            backend=backend,
+            differentiation_rule=differentiation_rule,
+        )
+
+    raise_error(
+        NotImplementedError,
+        "Only tensorflow automatic differentiation is supported at this moment.",
+    )
+
+
+def expectation_with_tf(
+    observable: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    initial_state: Optional[Union[List, qibo.Circuit]] = None,
+    nshots: int = 1000,
+    backend: str = "qibojit",
+    differentiation_rule: Optional[callable] = parameter_shift,
+):
```

Comment:
Since you already have the defaults in `expectation`, and that's the user facing one, I'd keep this one defaultless (to have a single source of truth).
```suggestion
def expectation_with_tf(
    observable: qibo.hamiltonians.Hamiltonian,
    circuit: qibo.Circuit,
    initial_state: Optional[Union[List, qibo.Circuit]],
    nshots: int,
    backend: str,
    differentiation_rule: Optional[callable],
):
```